GEM and ME0 segments 81 x

jshlee opened the pull request:
added GEMSegment in dataFormat
added GEMSegment builder
added common muon segment fitting
added GEMSegment to run in run3_GEM
added new ME0SegAlgo to use the common muon segment fitting

jshlee committed:
GEMsegments added
added common muon seg fit code

jshlee committed:
update RecoLocalMuon cfg

jshlee committed:
update RecoLocalMuon cfg

jshlee committed:
revert BuildFile.xml changes

cmsbuild commented:
A new Pull Request was created by  @jshlee (Jason Lee) for CMSSW_8_1_X.

It involves the following packages:

DataFormats/GEMRecHit
RecoLocalMuon/Configuration
RecoLocalMuon/GEMRecHit
RecoLocalMuon/GEMSegment


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@battibass, @abbiendi, @jhgoh, @bellan, @HuguesBrun, @trocino this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

slava77 commented:
@jshlee please provide more details on the proposed code.
A link to a talk in muon or related pog would help.

slava77 commented on DataFormats/GEMRecHit/src/classes_def.xml:
```diff
@@ -36,6 +36,15 @@
   <class name="GEMCSCSegmentRef" splitLevel="0"/>
 
+  <class name="GEMSegment" ClassVersion="-1">
+  <version ClassVersion="-1" checksum="1513163288"/>
```
why "-1" ?
for sensible types we expect something with a class version starting from 2 (maybe even better from 3)

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.cc:
```diff
@@ -0,0 +1,499 @@
+  }  
+
+  // edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::isGoodToMerge] :: oldChain ::";
+  for(size_t iRH_old = 0;iRH_old<oldChain.size();++iRH_old){
```
if this extra debug  ``` edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::is ...``` is not needed, please remove it from the commits

... same for other places

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 18ab2ef8db5d614fe1d0cfd8948325acbdff8554
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/12857/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/12857/summary.html

jshlee commented:
@slava77 - here is the link to the talk i gave on this https://indico.cern.ch/event/523877/contributions/2153651/attachments/1266133/1874312/gem_20160429.pdf

jshlee committed:
GEMSegment changed ClassVersion to 3
removed useless LogVerbatim statements

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

slava77 commented on DataFormats/GEMRecHit/interface/GEMSegment.h:
```diff
@@ -0,0 +1,119 @@
+
+    virtual int dimension() const { return 4; }
+
+    virtual int degreesOfFreedom() const { return 2*nRecHits() - 4;}	 
```
can this reallistically be anything other than 0?

jshlee replied:
for ge21, it can have more than 2 hits

slava77 replied:
this is for the wrong geometry which has 4 layers (short and long superchambers), right?
IIRC, only one long superchamber will be made

jshlee replied:
i thought that there could be more layers... I'll check this.

slava77 replied:
BTW, this was more of a curiosity question. No need to change the code here at the moment.

slava77 commented on DataFormats/GEMRecHit/interface/GEMSegment.h:
```diff
@@ -0,0 +1,119 @@
+    // and always choose Layer 1 in general
+      DetId detid = geographicalId(); 
+      GEMDetId rollid = GEMDetId(detid);
+      GEMDetId chamid = rollid.chamberId();
```
Is this commented out block really needed?

jshlee replied:
its not needed, will remove

slava77 commented on DataFormats/GEMRecHit/interface/GEMSegment.h:
```diff
@@ -0,0 +1,119 @@
+
+    float time()    const { return theTimeValue; }
+    float timeErr() const { return theTimeUncrt; }
+    float BunchX()  const { return theBX; }
```
please start with lower case (CMSSW style)

jshlee replied:
yes will do

slava77 commented on DataFormats/GEMRecHit/src/GEMSegment.cc:
```diff
@@ -0,0 +1,147 @@
+ *  \author Piet Verwilligen
+ */
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <DataFormats/GEMRecHit/interface/GEMSegment.h>
```
please use ``` " " ``` for includes of CMSSW files

slava77 commented on DataFormats/GEMRecHit/interface/GEMSegment.h:
```diff
@@ -0,0 +1,119 @@
+	       LocalVector direction, AlgebraicSymMatrix errors, double chi2, float bx);
+
+    GEMSegment(const std::vector<const GEMRecHit*>& proto_segment, LocalPoint origin, 
+	      LocalVector direction, AlgebraicSymMatrix errors, double chi2, double time, double timeErr);
```
please pass all large objects (LocalPoint, LocalVector, AlgebraicSymMatrix) by const reference

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.cc:
```diff
@@ -0,0 +1,479 @@
+    // First added hit in seed defines the mean to which the next hit is compared
+    // for this seed.
+    // running_meanX.push_back( rechits[i]->localPosition().x() );
+    // running_meanY.push_back( rechits[i]->localPosition().y() );
```
is this commented out code needed?
please cleanup

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.cc:
```diff
@@ -0,0 +1,479 @@
+    // running_meanY.push_back( rechits[i]->localPosition().y() );
+    GEMDetId rhID                  = rechits[i]->gemId();
+    const GEMEtaPartition * rhEP   = (theEnsemble.second.find(rhID.rawId()))->second;
+    if(!rhEP) throw cms::Exception("GEMEtaPartition not found") << "Corresponding GEMEtaPartition to GEMDetId: "<<rhID<<" not found in the GEMEnsemble";
```
I suppose this can only indicate there is a problem in configuration or there is a bug in the code.
If not, there should be some other way to handle the run time error.

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.cc:
```diff
@@ -0,0 +1,479 @@
+  edm::LogVerbatim("GEMSegAlgo") << "[GEMSegAlgo::buildSegments] will now wrap fit info in GEMSegment dataformat";
+  // GEMSegment tmp(proto_segment, protoIntercept, protoDirection, protoErrors, protoChi2);
+  GEMSegment tmp(proto_segment, protoIntercept, protoDirection, protoErrors, protoChi2, bx);
+  // GEMSegment tmp(proto_segment, protoIntercept, protoDirection, protoErrors, protoChi2, averageTime, timeUncrt);
```
I see how L470 may be needed. 
Why do we need L468 commented out alt construction?

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.h:
```diff
@@ -0,0 +1,86 @@
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.h>
+#include <DataFormats/GEMRecHit/interface/GEMRecHit.h>
```
please use ``` " " ``` for CMSSW includes

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.h:
```diff
@@ -0,0 +1,86 @@
+
+class MuonSegFit;
+
+class GEMSegAlgo : public GEMSegmentAlgorithm {
```
this should be more creative and more descriptive in the naming (if anything, GEMSegAlgo could be a typedef to GEMSegmentAlgorithm, not a different implementation)

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.h:
```diff
@@ -0,0 +1,86 @@
+  const std::string myName; 
+
+  // input from .cfi file
+ private:
```
there is no need to list ``` private ``` multiple times

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.h:
```diff
@@ -0,0 +1,86 @@
+  GEMEnsemble theEnsemble;
+  GEMDetId    theChamberId;
+
+  static constexpr float running_max=999999.;
```
maybe ``` std::numeric_limits<float>::max ``` ?

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.h:
```diff
@@ -0,0 +1,40 @@
+#include <DataFormats/GEMRecHit/interface/GEMSegment.h>
+#include <Geometry/GEMGeometry/interface/GEMSuperChamber.h>
+// #include <Geometry/GEMGeometry/interface/GEMChamber.h>
+#include <Geometry/GEMGeometry/interface/GEMEtaPartition.h>
```
Please clean up the includes

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc:
```diff
@@ -0,0 +1,123 @@
+    std::vector<GEMSegment> segv = algo->run(ensemble, gemRecHits);
+    GEMDetId mid(enIt->first);
+    edm::LogVerbatim("GEMSegmentBuilder") << "[GEMSegmentBuilder::build] found " << segv.size() << " segments in GEM Super Chamber " << mid;
+    edm::LogVerbatim("GEMSegmentBuilder") << "[GEMSegmentBuilder::build] -----------------------------------------------------------------------------"; 
```
It looks like this will be called very frequently per event.
This is too costly. Please change to LogTrace or hide behind a configurable ``` if (debug ) ``` or similar

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.h:
```diff
@@ -0,0 +1,52 @@
+
+    std::string algoName;
+    edm::ParameterSet segAlgoPSet;
+    GEMSegmentAlgorithm* algo;
```
make this unique_ptr ? .. unless the factory owns created algos and keeps track of destruction.

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentProducer.h:
```diff
@@ -0,0 +1,34 @@
+#ifndef GEMRecHit_GEMSegmentProducer_h
+#define GEMRecHit_GEMSegmentProducer_h
```
contents of this header file can be fully included in .cc and .h is not needed.

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentProducer.h:
```diff
@@ -0,0 +1,34 @@
+    /// Destructor
+    ~GEMSegmentProducer();
+    /// Produce the GEMSegment collection
+    virtual void produce(edm::Event&, const edm::EventSetup&);
```
add ``` override ```

slava77 commented on RecoLocalMuon/GEMSegment/plugins/ME0SegAlgo.cc:
```diff
@@ -0,0 +1,396 @@
+ *  \authors: Marcello Maggi, Jason Lee
+ */
+ 
+#include "ME0SegAlgo.h"
```
please edit the PR title to mention ME0 segment algo as well.

slava77 commented on RecoLocalMuon/GEMSegment/plugins/ME0SegAlgo.cc:
```diff
@@ -0,0 +1,396 @@
+/* Constructor
+ *
+ */
+ME0SegAlgo::ME0SegAlgo(const edm::ParameterSet& ps) : ME0SegmentAlgorithm(ps), myName("ME0SegAlgo")
```
the implementation looks like a copy of ME0SegAlgoMM with the same interface and ~95% the same implementation.
Please edit ME0SegAlgoMM instead, to deliver the slight change in the algorithm with a configuration option.

jshlee replied:
i would rather delete the old ME0SegAlgoMM and ME0SegFit since they are exactly the same

slava77 commented on RecoLocalMuon/GEMSegment/plugins/MuonSegFit.cc:
```diff
@@ -0,0 +1,470 @@
+  //       x = uz + c1 
+  //       y = vz + c2
+
+  edm::LogVerbatim("MuonSegFit") << "[MuonSegFit:fit2]-------------------------------------";
```
This will be called many times per event, which is rather costly.
Please switch to LogDebug or hide behind ``` if (debug ) ```

slava77 commented on RecoLocalMuon/GEMSegment/plugins/SealModule.cc:
```diff
@@ -4,6 +4,15 @@
+#include <RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilderPluginFactory.h>
+#include <RecoLocalMuon/GEMSegment/plugins/GEMSegAlgo.h>
+
+DEFINE_FWK_MODULE(GEMSegmentProducer);
```
``` DEFINE_FWK_MODULE ``` can be done in the GEMSegmentProducer.cc, which is a more common way since ~2010

slava77 commented:
I'm done with the review. The following issues stand out:
- excessive use of LogVerbatim (a verbatim version of LogInfo, which is compiled, executed, but is typically not displayed by the MessageLogger)
- ME0SegAlgo can replace or be replaced by ME0SegAlgoMM
- ``` class GEMSegAlgo : public GEMSegmentAlgorithm ``` better be more descriptive in naming

jshlee committed:
slava's suggestions

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

slava77 commented on DataFormats/GEMRecHit/interface/GEMSegment.h:
```diff
@@ -0,0 +1,99 @@
+	       const LocalVector direction, const AlgebraicSymMatrix errors, double chi2);
+
+    GEMSegment(const std::vector<const GEMRecHit*>& proto_segment, const LocalPoint origin, 
+	       const LocalVector direction, const AlgebraicSymMatrix errors, double chi2, float bx);
```
these are now const, but are still by value

slava77 commented:
Jason, please let me know when you are done with updates.
I understood that ME0SegAlgoMM vs ME0SegAlgo will be resolved here, in this PR.

jshlee committed:
MuonSegFit - cleaned up LogVerbatim

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee committed:
removed ME0SegAlgoMM and ME0SegFit

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
hi @slava77 - i've made the updates
 * i've moved the LogVerbatim to EDM_ML_DEBUG
 * ME0SegAlgoMM and ME0SegFit removed
 * class GEMSegAlgo : public GEMSegmentAlgorithm not sure what to do here...

do you want ME0SegmentProducer to be changed like GEMSegmentProducer??

slava77 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.h:
```diff
@@ -0,0 +1,39 @@
+#include <map>
+#include <vector>
+
+class GEMSegmentAlgorithm {
```
maybe rename this to AbsGEMSegmentAlgorithm to indicate it's abstract; GEMSegmentAlgorithmBase would also work.

slava77 replied:
... and in this case the GEMSegAlgo should become GEMSegmentAlgorithm for the lack of a better specification, since it's the first derived version of this abstract type

jshlee replied:
yep ok. shall I do the same for ME0 too?

slava77 replied:
sure, it would be nice.

slava77 commented:
```
do you want ME0SegmentProducer to be changed like GEMSegmentProducer??
```
do you mean the .h merged into .cc or something else?

jshlee committed:
GEMSegment - constructor input changed to const reference

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
the .h merged to .cc and the DEFINE_FWK_MODULE moved to the .cc

jshlee committed:
GEMSegmentAlgorithm -> GEMSegmentAlgorithmBase
GEMSegAlgo -> GEMSegmentAlgorithm

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
GEMSegmentAlgorithm -> GEMSegmentAlgorithmBase
GEMSegAlgo -> GEMSegmentAlgorithm
has been done.

I think the only things left for now are ME0SegmentProducer, ME0SegmentAlgorithm and ME0SegAlgo

jshlee committed:
ME0Segment - applying slava's gemSegment comments
ME0SegmentAlgorithm -> ME0SegmentAlgorithmBase
ME0SegAlgo -> ME0SegmentAlgorithm
ME0Segment change constructor input to const reference
ME0SegmentProducer.h merged to ME0SegmentProducer.cc
ME0SegmentBuilder change ME0SegmentAlgorithm to unique_ptr
bug fix - need to change SegmentAlgorithm to SegmentAlgorithmBase in
SegmentBuilder
change include <> to ""

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
updated ME0Segment - applying slava's gemSegment comments 
ME0SegmentAlgorithm -> ME0SegmentAlgorithmBase
ME0SegAlgo -> ME0SegmentAlgorithm
ME0Segment change constructor input to const reference
ME0SegmentProducer.h merged to ME0SegmentProducer.cc
ME0SegmentBuilder change ME0SegmentAlgorithm to unique_ptr
bug fix - need to change SegmentAlgorithm to SegmentAlgorithmBase in SegmentBuilder
change include <> to "" in RecoLocalMuon/GEMSegment and DataFormats/GEMRecHit

@slava77 - I think all the changes are done now, can you test please

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 183992dd059e1807f849bd9a34ee6b5c3fcd5ff4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/12873/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/12873/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/1440	183992d
- code changes are in line with the description and also include follow up to the code review as expected. The GEM segments provide a good reduction of (mostly neutron) background noise through correlation of hits on two layers of GEM superchambers
- jenkins tests pass and comparisons show no differences (only 2023 workflows are affected, which are not in jenkins set of tests)
- tested locally using wf 11024: gemSegments show up in the outputs
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->       750.1        750     NEWO   0.23     GEMDetIdGEMSegmentsOwnedRangeMap_gemSegments__RECO.
-------------------------------------------------------------
   322534 ->      323283        749             0.2     ALL BRANCHES
```

CPU on this no-PU sample is small
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.20%         0.00 ms/ev ->         2.42 ms/ev gemSegments
```

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

davidlange6 commented on DataFormats/GEMRecHit/src/classes_def.xml:
```diff
@@ -36,6 +36,15 @@
+  </class>
+  <class name="std::vector<GEMSegment*>" splitLevel="0"/>
+  <class name="edm::OwnVector<GEMSegment,edm::ClonePolicy<GEMSegment> >" splitLevel="0"/>
+  <class name="edm::RangeMap<GEMDetId,edm::OwnVector<GEMSegment,edm::ClonePolicy<GEMSegment> >,edm::ClonePolicy<GEMSegment> >" splitLevel="0"/>
```
@jshlee - why override the default splitLevel for these objects?

slava77 replied:
it's a copy-paste, used similarly for all other muon hit types, since long time ago.

davidlange6 replied:
ok, thanks - i guess the reason is lost in time..

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc:
```diff
@@ -0,0 +1,472 @@
+      // hits on the same layer should not be allowed ==> if abs(layer_new - layer_old) > 0 is ok. if = 0 is false
+      // allow only change true --> false. once false, leave it false
+      if(layerRequirementOK==true) { 
+	layerRequirementOK = abs(layer_new[jRH_new]-layer_old[jRH_old]) > 0;
```
layer_new[jRH_new] != layer_old[jRH_old] would be much faster

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc:
```diff
@@ -0,0 +1,472 @@
+
+bool GEMSegmentAlgorithm::isGoodToMerge(const EnsembleHitContainer& newChain, const EnsembleHitContainer& oldChain) {
+
+  std::vector<float> phi_new, eta_new, phi_old, eta_old;
```
call reserve on each of these vectors - you know the size of each of them

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc:
```diff
@@ -0,0 +1,124 @@
+    // save current GEMRecHit in vector
+    pp.push_back(it2->clone());
+    // assign updated vector of GEMRecHits to reference id
+    ensembleRH[id.rawId()]=pp;
```
Setting aside that its not good practice to have a map of vectors, why not ensembleRH[id.rawID()].push_back(it2->clone()) instead of these last three lines

jshlee committed:
david's comments
SegmentAlgorithm - reserve for vectors
SegmentBuilder - ensembleRH[id.rawId()].push_back(it2->clone())

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
@davidlange6 - I've updated both gem and me0 with your comments.
For overriding the default splitLevel, do you want this to be changed???

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/14400	fb28951
- updated as suggested by @davidlange6 
- local tests with 11024 show no difference in output product size and there may be a  reduction in CPU for gemSegments, although 0.07 ms/event  is well within uncertainties

```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   -0.029528      -0.01%         2.42 ms/ev ->         2.35 ms/ev gemSegments
```

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs after it passes the integration tests. This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

cmsbuild commented:
+1
Tested at: fb28951cc698204986df776a67e83ea5f0d5ae80
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/12917/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/12917/summary.html

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc:
```diff
@@ -0,0 +1,474 @@
+      // clear the buffer for the subset of segments:
+      segments_temp.clear();
+      // build the subset of segments:
+      segments_temp = this->buildSegments( (*sub_rechits) );
```
this would have 3x less allocations if buildSegments was instead

void buildSegments(const EnsembleHitContainer& rechits, std::vector<GEMSegment> &gemsegs);

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc:
```diff
@@ -0,0 +1,474 @@
+    
+    GEMRecHit *newRH = (*rh)->clone();
+    newRH->setPosition(lp);
+    muonRecHits.push_back(newRH);
```
@jshlee - how are these  pointers deleted?

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/ME0SegmentAlgorithm.cc:
```diff
@@ -44,21 +44,21 @@ ME0SegAlgoMM::ME0SegAlgoMM(const edm::ParameterSet& ps) : ME0SegmentAlgorithm(ps
-std::vector<ME0Segment> ME0SegAlgoMM::run(const ME0Ensemble& ensemble, const EnsembleHitContainer& rechits) {
+std::vector<ME0Segment> ME0SegmentAlgorithm::run(const ME0Ensemble& ensemble, const EnsembleHitContainer& rechits) {
 
   theEnsemble = ensemble;
```
this is a very heavy weight object to copy and then not change (as far as I see).Could you instead pass it down to the routine that needs it as a const reference?

jshlee replied:
@davidlange6 - I've done all the suggestions apart from this.
Do you want me to remove the global variable theEnsemble and just pass it to the functions that use this object?

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc:
```diff
@@ -0,0 +1,474 @@
+  for(unsigned int i = 0; i < rechits.size(); ++i) {
+    temp.clear();
+    temp.push_back(rechits[i]);
+    seeds.push_back(temp);
```
how about
ProtoSegments seeds();
seeds.reserve(rechits.size());
std::vector<bool> usedCluster(rechits.size(),false);
for ( unsigned int i=0; i<rechits.size(); ++i)
    seeds.push_back(EnsembleHitContainer(1,rechits[i]));

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc:
```diff
@@ -0,0 +1,474 @@
+  std::vector<int> bx_new, bx_old;
+  bx_new.reserve(newChain.size()); bx_old.reserve(oldChain.size());
+
+  for(size_t iRH_new = 0;iRH_new<newChain.size();++iRH_new){
```
I belatedly realize that all of the "new" variable calculations could be delayed until the loop starting on 342 (which would mean only the ones used before the function exits would be computed.

davidlange6 commented on RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc:
```diff
@@ -0,0 +1,474 @@
+      if(layerRequirementOK==true) { 
+	layerRequirementOK = layer_new[jRH_new] != layer_old[jRH_old];
+      }
+
```
I believe you can add

if (!layerRequirementOK) return false;

to save some calculations.

davidlange6 replied:
or better just remove layerRequirementOK completely and use

if ( layer_new[jRH_new] == layer_old[jRH_old] ) return false;

jshlee committed:
SegmentAlgorithm - david's comments

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

slava77 commented:
There are two more unaddressed comments.
@davidlange6 do you expect more comments from your side? It wasn't clear if you were done with the review.

jshlee commented:
I thought i addressed all the comments...
Can you remind me which ones I have left out.

slava77 commented:
The ones still displayed in this page
https://github.com/cms-sw/cmssw/pull/14400#discussion-diff-63152495
https://github.com/cms-sw/cmssw/pull/14400#discussion-diff-63154671
are either not addressed or were addressed in different context lines.
I didn't look in detail when I asked.

jshlee commented:
@slava77 - these two have been addressed.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 512fe8fac77dc4e1790ee14667432dce4fa57941
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13002/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13002/summary.html

slava77 commented:
I was looking at the latest code change effect in workflow 11024 (https://github.com/cms-sw/cmssw/compare/fb28951...512fe8fac77dc4e1790ee14667432dce4fa57941)
- timing changes in a good direction, reducing CPU time in me0Segments by x4 and  gemSegments by x2:
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   -1.267504      -7.52%       120.65 ms/ev ->        27.05 ms/ev me0Segments
   -0.648161      -0.09%         2.35 ms/ev ->         1.20 ms/ev gemSegments
```
- however, the event content changes:
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
    749.1 ->      1077.9        329     36.0   0.10     GEMDetIdGEMSegmentsOwnedRangeMap_gemSegments__RECO.
# number of segments changes from 11 to 27
    612.7 ->       512.2       -100    -17.9  -0.03     ME0DetIdME0SegmentsOwnedRangeMap_me0Segments__RECO.
# number of segments changes from 9 to 6
-------------------------------------------------------------
   323281 ->      323511        230             0.1     ALL BRANCHES
```

@jshlee 
did you check that the latest changes are correct algorithmically?
Please confirm this is what you need; otherwise some debugging will be necessary.

jshlee committed:
Merged GEMSegments-81X from repository jshlee

jshlee committed:
bugfix for gemsegments

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
@slava77 - i've fixed the gemSegments so they should be the same as before.
the me0Segments should also be fixed, but i still see a difference. this new update removes bad me0Segments.
btw, how can i run the test you do?

slava77 commented:
@cmsbuild please test

@jshlee 
I was just checking the branch sizes before and after 
using https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_output_products
compareProducts.sh script
and after seeing them different just manually checked the sizes.
The next step would be to add it to validate.C, used for FWLite-based validation, and run comparison as described here
https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_outputs_FWLite_style

cmsbuild commented:
The tests are being triggered in jenkins.

jshlee committed:
Merged GEMSegments-81X from repository jshlee

cmsbuild commented:
+1
Tested at: 0db2de3271f79962f08729a570fc67cb364d1245
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13061/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/2dfdf9494ece7dfc99b6c986bb4312649d6167b7
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13061/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13061/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13061/summary.html

slava77 commented:
@jshlee 
regarding 
https://github.com/cms-sw/cmssw/pull/14400#issuecomment-220060425
I don't see any substantive change in ME0 segments from the last commits
https://github.com/cms-sw/cmssw/compare/512fe8f...0db2de3271f79962f08729a570fc67cb364d1245?w=1
abs->std::abs for integers and ```>= ``` instead of ```> ``` for configuration cuts shouldn't have any substantive effect.
Maybe there's a missing commit (since you mention that me0Segments are fixed)?

I checked the output products in 11024.0 workflow and gemSegments indeed recover to the size they were right before the improvements done in https://github.com/cms-sw/cmssw/compare/fb28951...512fe8fac77dc4e1790ee14667432dce4fa57941
The me0Segments are unchanged still with a significant reduction in count as noted in 
https://github.com/cms-sw/cmssw/pull/14400#issuecomment-219864684

Jason, do you have any validation plots to show that the ME0 segment efficiency is still OK?
The speed improvement is very nice, but it looks like it changed behavior and probably broke the me0 segment reco.

jshlee committed:
Merged GEMSegments-81X from repository jshlee

slava77 commented:
@jshlee 
I compared on 100 events of 11021 (ten mu-e) to get more stats: the me0Segments branch changes as follows, comparing with baseline (CMSSW_8_1_X_2016-05-10-2300) to baseline+thisPR
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
    750.1 ->       505.9       -244    -38.9  -0.10     ME0DetIdME0SegmentsOwnedRangeMap_me0Segments__RECO.
```
Interactive inspection shows that the number of me0Segments changes from 268 to 188, about 30% dop.

If I'm not mistaken, this change comes from the seemingly technical improvements done in https://github.com/cms-sw/cmssw/compare/fb28951...512fe8fac77dc4e1790ee14667432dce4fa57941
Is this expected or to be fixed?

jshlee committed:
revert back to old me0segments

jshlee commented:
@slava77 - i found out where the change was happening and changed me0Segments back to the old numbers.

cmsbuild commented:
Pull request #14400 was updated. @ghellwig, @ianna, @cerminar, @Martin-Grunewald, @rekovic, @vanbesien, @perrotta, @civanch, @cmsbuild, @fwyzard, @davidlange6, @smuzaffar, @Dr15Jones, @cvuosalo, @mdhildreth, @deguio, @slava77, @ggovi, @mmusich, @mulhearn, @dmitrijus, @franzoni can you please check and sign again.

jshlee committed:
revert back to old me0segments

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5477f474c78529d822ec299c36ca24d098addced
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13115/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13115/summary.html

slava77 commented on RecoLocalMuon/GEMSegment/plugins/ME0SegmentAlgorithm.cc:
```diff
@@ -319,15 +318,17 @@ void ME0SegmentAlgorithm::buildSegments(const ME0Ensemble& ensemble, const Ensem
     
     muonRecHits.push_back(newRH);    
   }
+  if (rechits.size() < minHitsPerSegment) return;
```
this could have stayed at the entry to the function: rechits is const and the output (me0segs ) are not modified until later.

jshlee replied:
should i change back to the start of the function? 
I thought it would be best to just keep it the same as the previous version.

slava77 replied:
Better move it back up.

slava77 commented on RecoLocalMuon/GEMSegment/plugins/ME0SegmentAlgorithm.cc:
```diff
@@ -319,15 +318,17 @@ void ME0SegmentAlgorithm::buildSegments(const ME0Ensemble& ensemble, const Ensem
   for (auto rh:muonRecHits) delete rh;
   // quit function if fit was not OK  
-  if(!goodfit) return;
+  //if(!goodfit) return;
```
this looks like the only change that would modify the results

jshlee committed:
ME0SegmentAlgorithm move minHitsPerSegment to top of function

jshlee committed:
Merge branch 'GEMSegments-81X' of https://github.com/jshlee/cmssw into GEMSegments-81X

jshlee committed:
ME0SegmentAlgorithm move minHitsPerSegment to top of function

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

slava77 commented:
@jshlee it looks like you have probably rebased (merge commit present)
If so, to which IB?

slava77 commented:
with a closer look, the base IB is still 810pre5, no actual rebase.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: cbd975836811108a284a27dec5f0c58186cd8ef1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13139/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/f7c9fc1b25a9ac3984b1ca74ce4f5c1fa542f649
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13139/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13139/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13139/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/14400	 	cbd9758
- code changes are in line with the description and have addressed the follow up code review
- jenkins tests pass and comparisons with baseline show no differences (the 2019/phase2 workflows are not in the set of tests)
- local test on 100 events in 11021.0 (10mue up to local reco in 2023 setup) show that gemSegments appear in the outputs; ME0 segment size is the same while CPU is reduced substantially
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->      1343.3       1343     NEWO   0.54     GEMDetIdGEMSegmentsOwnedRangeMap_gemSegments__RECO.
    749.4 ->       746.4         -3     -0.4  -0.00     ME0DetIdME0SegmentsOwnedRangeMap_me0Segments__RECO.
# Manual check: 268 segmenss -> 268 : OK
-------------------------------------------------------------
   247423 ->      248764       1341             0.5     ALL BRANCHES

```
and timing
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.29%         0.00 ms/ev ->         1.25 ms/ev gemSegments
   -1.337589     -19.52%       103.65 ms/ev ->        20.57 ms/ev me0Segments
```

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

jshlee commented:
@kpedro88 - gemsegment PR for pre6

kpedro88 commented:
@davidlange6 can you merge this one also?

davidlange6 commented:
it seems there are still some outstanding questions - at least the one about pointer deletion is worth knowing the answer to soon.
@jshlee @kpedro88

slava77 commented:
@davidlange6 
github has hidden all your comments after recent edits.
Please enter them inline to better understand the remaining issues.

davidlange6 commented:
the reason I remembered is that I looked at the comments still inline (indeed, anything where the comment is gone is also lost from my memory). You don't see any old comments?

jshlee commented:
@davidlange6 - I deleted the pointers in https://github.com/cms-sw/cmssw/pull/14400/files#diff-09ca74bfd4f2fefcd97dae6f58ebb100R351
I'm pretty sure I took care of all the comments you made.

slava77 commented:
On 5/26/16 6:50 AM, David Lange wrote:
> the reason I remembered is that I looked at the comments still inline
> (indeed, anything where the comment is gone is also lost from my
> memory). You don't see any old comments?

indeed, I see them when I look at /files
https://github.com/cms-sw/cmssw/pull/14400/files


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14400#issuecomment-221876331>
>

kpedro88 commented:
Specifically these two from @davidlange6:
https://github.com/cms-sw/cmssw/pull/14400/files#diff-09ca74bfd4f2fefcd97dae6f58ebb100R130
https://github.com/cms-sw/cmssw/pull/14400/files#diff-09ca74bfd4f2fefcd97dae6f58ebb100R333

kpedro88 commented:
Ah, the second one about pointer deletion is indeed addressed by @jshlee at the line provided. You should check on the other one, though, about using reserve with a vector...

jshlee commented:
i'll fix this now

jshlee committed:
gem and me0 seg clusterHits updates

jshlee commented:
done. can you retest please

kpedro88 commented:
@cmsbuild please test

davidlange6 commented:
+1

ok - thanks - strange i didn’t figure it from github - i’ll merge this and further suggestions can come via another pr.

> On May 26, 2016, at 4:04 PM, Kevin Pedro <notifications@github.com> wrote:
> 
> Ah, the second one about pointer deletion is indeed addressed by @jshlee at the line provided. You should check on the other one, though, about using reserve with a vector...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #14400 was updated. @cmsbuild, @cvuosalo, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: a85670ce955c5c2c7b26c1a497dfe76cc3cf4cd6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13214/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14400/13214/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * 1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4+ALCAHARVD5

kpedro88 commented:
@slava77 can you sign again? Then it should be ready to merge

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/14400	a85670c
- since the last approval the code is updated to possibly more optimally fill vectors by reserving size and filling with implicit temporary. Code modified is for fairly light objects (effectively 1D vectors of basic types): no significant computational gain is expected.
- jenkins tests pass and show no differences (the changed code is not exercised)
- local check with 100 events in 11021.0 shows there are no changes in me0 and gem segments; comparison for timing change of these modules was not reliable (jobs ended up on different node kinds; scaling to orthogonal module time suggested an increase in CPU from the last changes)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request will be automatically merged.
